src/backend/gl/gl_common.c

```diff
@@ -623,13 +624,119 @@ bool gl_kernel_blur(backend_t *base, double opacity, void *ctx, const rect_t *ex
 	return true;
 }
 
-bool gl_dual_kawase_blur(backend_t *base, double opacity attr_unused, void *ctx,
-                         const rect_t *extent attr_unused, const int nrects attr_unused,
-                         const GLuint vao[2] attr_unused) {
-	auto bctx attr_unused = (struct gl_blur_context *)ctx;
-	auto gd attr_unused = (struct gl_data *)base;
+bool gl_dual_kawase_blur(backend_t *base, double opacity, void *ctx, const rect_t *extent,
+                         const int width, const int height, const int nrects,
+                         const GLuint vao[2]) {
+	auto bctx = (struct gl_blur_context *)ctx;
+	auto gd = (struct gl_data *)base;
+
+	int dst_y_screen_coord = gd->height - extent->y2,
+	    dst_y_fb_coord = bctx->fb_height - extent->y2;
+
+	// Reduce number of iterations until the last one renders at least 1px in both
+	// dimensions
+	int iterations = bctx->blur_texture_count;
+	while (((width / (1 << iterations)) < 1 || (height / (1 << iterations)) < 1) &&
```

Comment:
Same thing as `width >> iterations`, `height >> iterations`, I believe?

Reply:
Curious: does this mean smaller windows will get weaker blur?

Reply:
> Same thing as `width >> iterations`, `height >> iterations`, I believe?

Now that you say it. Should yield the same results, yes. ^^
And yes, we skip those iterations where we would render less than a full pixel in either width or height.

Reply:
Hmm, I wonder if it's going to look weird.

Not relevant to this PR, but I am thinking if we could always blur the whole screen, cache the result, and clip out the needed parts when rendering blurred background. This way we could get consistent blur strength, but I am not sure about performance.

Reply:
Windows have to be really tiny to notice. This was implemented mainly for performance reasons but you will get a noticeable change in the blur itself when you cross the threshold. We should be fine with leaving this one out (not sure about the OpenGL spec but my nvidia hardware doesn't give me any glitches when drawing less than a single pixel wide lines, probably because we already sample more due to `resize_width`/`resize_height` being added).

Regarding full-screen blur: Unless we are using `--transparent-clipping`, we would need to blur the whole screen multiple times (at least for each "stacking layer"). Not sure if that logic would be more performant than rendering each window separately. Restricting blur to xinerama-regions/xrandr-outputs should be easier though. Definitely stuff for later.

Reply:
> We should be fine with leaving this one out 

Keep it in, even without the check, the blur is going to be weaker anyway, right?

Reply:
> Keep it in, even without the check, the blur is going to be weaker anyway, right?

No, it should be consistent regardless of the window size.

I've done some testing and at the highest level (i.e. 5 iterations), the window has to be smaller than 32px in either direction for that to be noticable. With the check removed I could not produce any visual glitches with windows only a few pixels wide/tall. But I am pretty sure that is because we compensate with `resize_width` and `resize_height` anyway to have enough pixels outside the window for the last iteration.

I've implemented this check back in the old compton codebase without the expand/resize logic and where textures were allocated for each blur-call where less than 1px textures were a problem.

Reply:
I see, that makes sense then.